java/arcs/core/data/Recipe.kt

```diff
@@ -47,3 +48,31 @@ data class Recipe(
         }
     }
 }
+
+/** Translates a [Recipe] into a [Plan] */
+fun Recipe.toPlan() = Plan(
+    arcId = arcId,
+    particles = particles.map { it.toPlanParticle() }
+)
+
+/** Translates a [Recipe.Particle] into a [Plan.Particle] */
+fun Recipe.Particle.toPlanParticle() = Plan.Particle(
+    particleName = spec.name,
+    location = spec.location,
+    handles = handleConnections.associate { it.spec.name to it.toPlanHandleConnection() }
+)
+
+/** Translates a [Recipe.Particle.HandleConnection] into a [Plan.HandleConnection] */
+fun Recipe.Particle.HandleConnection.toPlanHandleConnection() = Plan.HandleConnection(
+    mode = spec.direction,
+    type = spec.type,
+    storageKey = handle.toStorageKey()
+    // TODO: Add TTL.
+)
+
+/** Translates a [Recipe.Handle] into a [StorageKey] */
+fun Recipe.Handle.toStorageKey() = when {
+    storageKey != null -> StorageKeyParser.parse(storageKey)
```

Comment:
Is it always the case that either `storageKey` or `capabilities` is set or nothing is set? If so add a `require` to check that?

Reply:
No, I don't think such guarantees can be made.

It is however the case that if storage key is provided, you need to use it. It if is not provided, you need to calculate it, and you calculate it based on capabilities and the name.

Reply:
Oh I see. In that case, what happens if both `capabilities` and `storageKey` are specified? It looks like this code prefers `storageKey` over `capabilities`. Is that the right thing to do?

Reply:
Capabilities are used to decide the storage medium. If storage key is already there, then most likely capability has already be "consulted" when calculating the storage key.

Reply:
I see. May be you want to document this so that some refactoring down the line does not change the order of the checks.

Reply:
I've added a test with the message documenting the priority.